### PR TITLE
Cheat n times 2

### LINKED ISSUE #1016: Design the single-call spanning cheatcode variants
This should include:
- warp
- roll
- prank
- spoof
We can thing of an alternative solution to foundry since it's not that obvious and might be confusing for users

### PR DESCRIPTION
**Pull request by drknzz**
Closes #1016 

## Introduced changes

Changes from previous version: [cheat-n-times-1](https://github.com/foundry-rs/starknet-foundry/commit/46b061a4b825fd1526080eb834ca375aec176b5f)

New changes:
1. [Calls -> TargetCalls](https://github.com/foundry-rs/starknet-foundry/commit/acc10dfd0b9c5ee3ed425c53aa23eae412846111)
2. [CheatTarget::Multiple documentation](https://github.com/foundry-rs/starknet-foundry/commit/5731f36c564c2c6b39d8f6e3db1e11ef23590721)
3. [mock_calls(amount) -> mock_call(n_times)](https://github.com/foundry-rs/starknet-foundry/commit/aaef50be1a9e5943d9245b62ca13fd36fc95db08)
4. [CheatTarget::All only with CheatSpan::Indefinite](https://github.com/foundry-rs/starknet-foundry/commit/47edaa6c8d8ca4cae9a3be1a2f61788c2fa5fd72)

## Checklist

- [x] Linked relevant issue
- [x] Updated relevant documentation
- [x] Added relevant tests
- [x] Performed self-review of the code
- [x] Added changes to `CHANGELOG.md`

### CODE REVIEW

File: crates/cheatnet/src/runtime_extensions/forge_runtime_extension/mod.rs [hunk 1]
```diff
@@ -78,6 +79,15 @@ impl BufferReaderExt for BufferReader<'_> {
             _ => unreachable!("Invalid CheatTarget variant"),
         })
     }
+
+    fn read_cheat_span(&mut self) -> BufferReadResult<CheatSpan> {
+        let cheat_span_variant = self.read_felt()?.to_u8();
+        Ok(match cheat_span_variant {
+            Some(0) => CheatSpan::Indefinite,
+            Some(1) => CheatSpan::Number(self.read_felt()?.to_usize().unwrap()),
+            _ => unreachable!("Invalid CheatSpan variant"),
```
Comment (piotmag769) [inline crates/cheatnet/src/runtime_extensions/forge_runtime_extension/mod.rs:88]: We should bubble errors instead there now https://github.com/foundry-rs/starknet-foundry/pull/1843/files#r1524983099

File: snforge_std/src/cheatcodes.cairo [hunk 2]
```diff
@@ -22,83 +29,116 @@ fn test_address() -> ContractAddress {
...
     target.serialize(ref inputs);
     cheatcode::<'stop_elect'>(inputs.span());
 }
 
-fn start_mock_call<T, impl TSerde: core::serde::Serde<T>, impl TDestruct: Destruct<T>>(
-    contract_address: ContractAddress, function_selector: felt252, ret_data: T
+fn mock_call<T, impl TSerde: core::serde::Serde<T>, impl TDestruct: Destruct<T>>(
+    contract_address: ContractAddress, function_selector: felt252, ret_data: T, n_times: u32
 ) {
+    assert!(n_times > 0, "n_times must be greater than 0");
```
Comment (piotmag769) [inline snforge_std/src/cheatcodes.cairo:115]: Cryptic error message

File: snforge_std/src/cheatcodes/tx_info.cairo [hunk 3]
```diff
@@ -44,20 +44,18 @@ impl TxInfoMockImpl of TxInfoMockTrait {
...
-    tx_info_mock.serialize(ref tx_info_serialized);
-
-    let mut inputs: Array<felt252> = array![];
-    inputs.append_span(cheat_target_serialized.span());
-    inputs.append_span(tx_info_serialized.span());
-
+fn spoof(target: CheatTarget, tx_info_mock: TxInfoMock, span: CheatSpan) {
+    let mut inputs = array![];
+    target.serialize(ref inputs);
+    span.serialize(ref inputs);
```
Comment (piotmag769) [inline snforge_std/src/cheatcodes/tx_info.cairo:50]: Missing taget + span validation

File: snforge_std/src/cheatcodes.cairo [hunk 4]
```diff
@@ -22,83 +29,116 @@ fn test_address() -> ContractAddress {
...
 
-fn start_prank(target: CheatTarget, caller_address: ContractAddress) {
+fn prank(target: CheatTarget, caller_address: ContractAddress, span: CheatSpan) {
+    validate_cheat_target_and_span(@target, @span);
+
     let mut inputs = array![];
     target.serialize(ref inputs);
+    span.serialize(ref inputs);
     inputs.append(caller_address.into());
     cheatcode::<'start_prank'>(inputs.span());
```
Comment (piotmag769) [inline snforge_std/src/cheatcodes.cairo:59]: I wonder if we shouldn't change the selectors to correct ones e.g. "prank", "roll" etc.

Comment (drknzz) [reply snforge_std/src/cheatcodes.cairo:59]: Could do

File: crates/cheatnet/src/runtime_extensions/forge_runtime_extension/mod.rs (hunk 1 above)
Comment (drknzz) [reply crates/cheatnet/src/runtime_extensions/forge_runtime_extension/mod.rs]: [Bubble CheatSpan parse error](https://github.com/foundry-rs/starknet-foundry/pull/1874/commits/c30c00326d08fd72c3abf3a6bbd71d905348e439)

File: snforge_std/src/cheatcodes.cairo (hunk 2 above)
Comment (drknzz) [reply snforge_std/src/cheatcodes.cairo]: [Change n_times == 0 error message](https://github.com/foundry-rs/starknet-foundry/pull/1874/commits/698fd081fcfdca9ff050a59e75ce9801824bca31)

File: snforge_std/src/cheatcodes/tx_info.cairo (hunk 3 above)
Comment (drknzz) [reply snforge_std/src/cheatcodes/tx_info.cairo]: [Add validation to spoof](https://github.com/foundry-rs/starknet-foundry/pull/1874/commits/6e6db5047be738fe979dd5fcca41a163ebf25339)

File: snforge_std/src/cheatcodes.cairo (hunk 4 above)
Comment (drknzz) [reply snforge_std/src/cheatcodes.cairo]: [Rename cheatcodes](https://github.com/foundry-rs/starknet-foundry/pull/1874/commits/df73cbd78e1cf5879bbfaba73a802e4d14be8a66)

**Review by piotmag769 — COMMENTED** [review]
The logic looks good, won't review the tests so let's wait for someone's else approve

**Pull request merged**